### DISCUSSION TITLE
Docs-as-Code functionality in Budibase

### DISCUSSION THREAD

**Discussion by gerhardmpl** [Budibase/budibase#6565] [Ideas]
It would be great to have the option to write documentation (concepts, ideas, decisions or comments / hints to myself or others) using the same tool one uses to create an App in Budibase.

Adding a long text (or markdown) field to the components setting bar or a new documentation component of that kind would allow to write down notes and generate a documentation based on the content (e.g. piping the content through a tool like docToolchain/AsciiDoc).

Spoiler - No real-world experience with Docs-as-Code here, but I would love to be able to document my code within the App.

**Comment by melohagan** [comment]
Hi @gerhardmpl 

This seems interesting. 

What do you mean by:
> same tool one uses to create an App in Budibase

I'm struggling to visualise the idea and the use-case, but I also don't have much experience in the way of "Docs as Code".

**Comment by gerhardmpl** [comment]
Hi @melohagan and thanks for looking into my idea.

It is probably a limited use case, but I would appreciate a place to keep my notes and documentation on my Budibase Apps.

As you know, a good documentation is key to understand coding, business logic and design decisions or code changes of every application - and everybody hates it (at least developers usually do). Documentation as Code refers to writing documentation with the same tools that you use to write the code (same editor = Budibase UI). It simply means: Document everything in the code and not in a separate tool (like Wiki, Notes, Gira, Word) to make life easier for developers.

But how can you do that in a low-code / no-code environment where the main idea is to have no or only very little code? The best place I can imagine is the Budibase App Overview Section, Data Section and Screen Creation Section UI. 

A simple first step could be adding a new tile to the App Overview Section where a (markdown) long text can be edited (in a modal window) and a preview of the first lines is shown (see mockup).

![grafik](https://user-images.githubusercontent.com/40722408/177810392-fdd4e7ed-a496-4f68-b198-c251976585be.png)

A next step could be adding a new element for the Screen Section UI where a (markdown) long text can be edited in a modal window or the elements settings bar (see next mockup). The element would not be available on the screen.

![grafik](https://user-images.githubusercontent.com/40722408/177810804-c70a754a-d50d-4390-ac58-54f81971ad67.png)

A future step could be to hierarchically read and combine all document elements to an overall documentation (with structure thanks to the markdown syntax), but I agree that might be to much.

Edit: it's markdown, not markup and one concern: would that bloat the apps size, making it slow loading?

**Reply by melohagan** [reply]
Thanks very much for the detailed explanation @gerhardmpl !

In hindsight I've actually used Swagger before which might fall into this category. Anyway, this is a really cool idea. 

Instead of a documentation component, it might be nicer to have a button for adding documentation within the screen settings, just to avoid clutter in the component tree. 
Plus it's meta data about your components, not a component itself.

I agree writing good documentation is a hassle, but I don't imagine this will be worked on in the near future to be honest.

**Reply by gerhardmpl** [reply]
Hi @melohagan 

> I don't imagine this will be worked on in the near future

No problem, did not expect that. Just wanted to raise awareness, that a good no-code/low-code environment can benefit from a good documentation method from a user/developer/manager perspective.
